src/promise.ts

```diff
@@ -0,0 +1,260 @@
+const STATE = {
+  FULFILLED: 'fulfilled',
+  REJECTED: 'rejected',
+  PENDING: 'pending',
+};
+
+class MyPromise {
+  private thenCbs: Function[] = [];
+  private catchCbs: Function[] = [];
+  private state = STATE.PENDING;
+  private value: unknown;
+  private onSuccessBind = this.onSuccess.bind(this);
+  private onFailBind = this.onFail.bind(this);
+
+  constructor(cb: { (resolve: Function, reject: Function): void }) {
+    try {
+      cb(this.onSuccessBind, this.onFailBind);
+    } catch (error) {
+      if (error instanceof Error) {
+        this.onFail(error);
+      }
+    }
+  }
+
+  private runCallbacks(): void {
+    if (this.state === STATE.FULFILLED) {
+      this.thenCbs.forEach((callback: Function) => {
+        callback(this.value);
+      });
+
+      this.thenCbs = [];
+    }
+
+    if (this.state === STATE.REJECTED) {
+      this.catchCbs.forEach((callback: Function) => {
+        callback(this.value);
+      });
+
+      this.catchCbs = [];
+    }
+  }
+
+  private onSuccess(value: unknown): void {
+    queueMicrotask(() => {
+      if (this.state !== STATE.PENDING) return;
+      if (value instanceof MyPromise) {
+        value.then(this.onSuccessBind, this.onFailBind);
+        return;
+      }
+
+      this.value = value;
+      this.state = STATE.FULFILLED;
+      this.runCallbacks();
+    });
+  }
+
+  private onFail(value: unknown): void {
+    queueMicrotask(() => {
+      if (this.state !== STATE.PENDING) return;
+
+      if (value instanceof MyPromise) {
+        value.then(this.onSuccessBind, this.onFailBind);
+        return;
+      }
+
+      if (this.catchCbs.length === 0 && typeof value === 'string') {
+        throw new UncaughtPromiseError(value);
+      }
+
+      this.value = value;
+      this.state = STATE.REJECTED;
+      this.runCallbacks();
+    });
+  }
+
+  then(
+    thenCb?: Function | null | undefined,
```

Comment:
```suggestion
    thenCb?: Function | null,
```
If you are using an optional param, you don't need to describe an undefined type